### PR TITLE
Fixes #36969 - return the actual exit status of registration command

### PR DESCRIPTION
**Pull request by nofaralfasi**

### CODE REVIEW

**Comment by nofaralfasi** [comment]
This change will provide the correct exit status for a failing registration command. 
However, it doesn't address the problem of the curl command failure because the bash command at the end of the pipeline (`| bash`) is executed regardless of the curl command's outcome. The pipeline's exit status depends on the last command (bash), which returns 0 even if the curl command fails.

@stejskalleos Any suggestions on how to fix this issue? I was thinking we could save the curl command output in a temporary file and execute it only in the case of a successful curl exit status.

**Comment by nofaralfasi** [comment]
Modified the curl command format to accurately reflect its exit status in case of a failure.

**Comment by nofaralfasi** [comment]
@ShimShtein can you please review?

File: app/controllers/concerns/foreman/controller/registration_commands.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@ module Foreman::Controller::RegistrationCommands
 
   def command
     args_query = "?#{registration_args.to_query}"
-    "curl -sS #{insecure} '#{registration_url(@smart_proxy)}#{args_query if args_query != '?'}' #{command_headers} | bash"
+    "output=$(curl -sS #{insecure} '#{registration_url(@smart_proxy)}#{args_query if args_query != '?'}' #{command_headers}) && echo \"$output\" | bash"
```
Comment (ShimShtein) [inline app/controllers/concerns/foreman/controller/registration_commands.rb]: > I was thinking we could save the curl command output in a temporary file and execute it only in the case of a successful curl exit status.

I would agree. It's better to save curl's output into a temp file, and then redirecting it into bash. This will be much cleaner and less prone to errors due to special symbols.

```suggestion
    "tmp=$(mktemp) && curl -o $tmp -sS #{insecure} '#{registration_url(@smart_proxy)}#{args_query if args_query != '?'}' #{command_headers} && cat #tmp | bash"
```

This should do the trick, but I didn't test it...

**Review by ShimShtein — CHANGES_REQUESTED** [review]

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 2]
```diff
@@ -98,7 +98,7 @@ fi
 <% end -%>
 
 register_host() {
-  curl --silent --show-error --cacert $SSL_CA_CERT --request POST <%= @registration_url %> \
+  register_tmp=$(mktemp) && curl -o $register_tmp --silent --show-error --cacert $SSL_CA_CERT --request POST <%= @registration_url %> \
```
Comment (ShimShtein) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: I think we should move the `register_tmp=$(mktemp)` outside the `register_host` and `register_katello_host`, to the same place we define the `SSL_CA_CERT` file. Then we can clean it up in the `cleanup_and_exit` function.
We can then pass the value of `register_temp` as a parameter to both `register` functions.

I think it would be much cleaner.

**Review by ShimShtein — CHANGES_REQUESTED** [review]

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 2 above)
Comment (ShimShtein) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: Also: we are mixing `&&` and `;`.
I think we can just write it as separate lines:
``` sh
curl && bash
EXIT_CODE=$?
#...
cleanup_and_exit $EXIT_CODE
```

**Review by ShimShtein — DISMISSED** [review]
Now it looks much better.
Didn't test it, but as far as the code goes - LGTM.

**Review by stejskalleos — CHANGES_REQUESTED** [review]
I've been looking into the templates:
* Global Registration
* Host_init_config_template

And I don't think the proposed fix with the tmp is the right way to go.

My testing:
Create file in `rails.root/public/file.sh`
```
#!/bin/sh

echo "YAY"!
exit 1

```

Run 
```
curl "http://foreman.example.com/file.sh" | bash; echo $?
=>
YAY!
1
```

So, the exit code **is passed** correctly, problem is somewhere else.

In the `host_init_config` template, we use `set +e` & `set -e`, but nothing like that
in the Global registration. I believe that's the problem, we should unify the workflows, catch the errors properly and return error code, not 0.

See https://github.com/theforeman/foreman/blob/develop/app/views/unattended/provisioning_templates/host_init_config/host_init_config_default.erb#L108

File: app/controllers/concerns/foreman/controller/registration_commands.rb [hunk 3]
```diff
@@ -5,7 +5,8 @@ module Foreman::Controller::RegistrationCommands
 
   def command
     args_query = "?#{registration_args.to_query}"
-    "curl -sS #{insecure} '#{registration_url(@smart_proxy)}#{args_query if args_query != '?'}' #{command_headers} | bash"
+    "tmp=$(mktemp) && curl -o $tmp -sS #{insecure} '#{registration_url(@smart_proxy)}#{args_query if args_query != '?'}' #{command_headers} && bash $tmp; CODE=$?; rm -f $tmp; (exit $CODE)
```
Comment (ShimShtein) [inline app/controllers/concerns/foreman/controller/registration_commands.rb]: @ekohl Need some bash expertise here: 
TL;DR: we want to fail the script if the `curl` command fails. The original solution `curl... | bash` doesn't work (it will always return status 0). 

Is this an acceptable solution, or is there a better one that we don't recognize?

Comment (stejskalleos) [reply app/controllers/concerns/foreman/controller/registration_commands.rb]: My two cents:
The solution in the PR is not wrong or bad IMO, it just adds a lot more complexity to the command, which I don't think is a good user experience.
I was able to achieve desired behavior with `set -o pipefail`

```
set -o pipefail
curl -sSk http://zz.redhat.com | bash;echo $?
=>
curl: (6) Could not resolve host: zz.redhat.com
6
```

Technically speaking, aren't we trying to fix behavior that is expected and correct? 
Maybe the fix is to configure user's CI pipeline with `set -o pipefail`

**Comment by maximiliankolb** [comment]
Ping @goarsna I believe this is related to https://github.com/theforeman/foreman/pull/9808

File: app/controllers/concerns/foreman/controller/registration_commands.rb (hunk 3 above)
Comment (evgeni) [reply app/controllers/concerns/foreman/controller/registration_commands.rb]: I'd say using `pipefail` is correct (it's also what Ansible suggests in their linter: https://ansible.readthedocs.io/projects/lint/rules/risky-shell-pipe/) and that's what we should incorporate in the generated command?

So make the PR here generate something like
```
"set -o pipefail && curl -sS #{insecure} '#{registration_url(@smart_proxy)}#{args_query if args_query != '?'}' #{command_headers} | bash"
```

Comment (stejskalleos) [reply app/controllers/concerns/foreman/controller/registration_commands.rb]: > that's what we should incorporate in the generated command?

IMHO we could just add it to the documentation.

Comment (evgeni) [reply app/controllers/concerns/foreman/controller/registration_commands.rb]: Why docs instead of here?

Comment (stejskalleos) [reply app/controllers/concerns/foreman/controller/registration_commands.rb]: to keep the command short as possible.
I know that with current length of the auth token it's not a big addition, but still ...

Comment (goarsna) [reply app/controllers/concerns/foreman/controller/registration_commands.rb]: I had a look at it. IMO `set -o pipefail` is definitely the way we should go. Personally I would also tend to extend the command as proposed by @evgeni. I have the feeling that adding it to the docs is a little bit unwieldy. Yes, it is only one extra bullet point in a list of 25 you have to follow when registering a host, but it's somehow interrupting the workflow between copying the generated curl command and pasting it on the host.

Comment (ShimShtein) [reply app/controllers/concerns/foreman/controller/registration_commands.rb]: I am also for automatic actions - the less we expect from the user, the less are chances for command failure.

**Comment by ares** [comment]
LGTM

**Review by stejskalleos — APPROVED** [review]
:green_apple: **LGTM**

Let's get this in!

Thanks @nofaralfasi and everyone for the reviews
